src/oni2-sparkle/stubs/winsparkle.c

```diff
@@ -0,0 +1,126 @@
+#include "config.h"
+
+#include <caml/alloc.h>
+#include <caml/callback.h>
+#include <caml/memory.h>
+#include <caml/mlvalues.h>
+#include <caml/threads.h>
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#ifdef USE_WIN_SPARKLE
+
+#include <windows.h>
+#include <shlwapi.h>
+
+#include "winsparkle.h"
+
+void oni2_WinSparkleCloseCallback() {
+  static const value *closeCallback;
+  if (closeCallback == NULL) {
+    closeCallback = caml_named_value("oni2_close");
+  }
+
+  if (closeCallback != NULL) {
+    caml_c_thread_register();
+    caml_acquire_runtime_system();
+    caml_callback(*closeCallback, Val_unit);
+    caml_release_runtime_system();
+  }
+}
+
+CAMLprim value oni2_SparkleInit() {
+  CAMLparam0();
+
+  char iniPath[MAX_PATH];
+
+  GetModuleFileName(NULL, iniPath, MAX_PATH);
+  PathRemoveFileSpec(iniPath);
+
+  strcat(iniPath, "\\Oni2.ini");
+  char version[16];
+  wchar_t versionWide[16];
+  
+
+  GetPrivateProfileString("Application", "Version", "", version, 16, iniPath);
+
+  mbstowcs(versionWide, version, 16);
+
+  win_sparkle_set_app_details(L"Outrun Labs LLC", L"Onivim 2", versionWide);
+  win_sparkle_set_shutdown_request_callback(oni2_WinSparkleCloseCallback);
+  win_sparkle_init();
+
+  CAMLreturn(Val_unit);
+}
+
+// On WinSparkle, there is no instance -- the library just uses global variables.
+// So here we just return a basic `unit`
+CAMLprim value oni2_SparkleGetSharedInstance() {
+  CAMLparam0();
+
+  CAMLreturn(Val_unit);
+}
+
+// This function doesn't make too much sense since we lose the OOP aspect on
+// Windows, so it's basically a noop.
+CAMLprim value oni2_SparkleDebugToString(value vData) {
+  CAMLparam1(vData);
+  CAMLlocal1(vStr);
+
+  vStr = caml_copy_string("Unimplemented");
+
+  CAMLreturn(vStr);
+}
+
+// Again, this function doesn't make much sense on Windows, so it's a noop.
+CAMLprim value oni2_SparkleDebugLog(value vData) {
+  CAMLparam1(vData);
+  
+  CAMLreturn(Val_unit);
+}
+
+CAMLprim value oni2_SparkleSetFeedURL(value vUpdater, value vUrlStr) {
+  CAMLparam2(vUpdater, vUrlStr);
+  static int initialized = 0;
+
+
+  win_sparkle_set_appcast_url(String_val(vUrlStr));
```

Comment:
Does sparkle make a copy of this string? If not - we might need to. The GC can move strings around, if we just keep the pointer.

Reply:
You're right, it doesn't look like they make a copy of the string: 
https://github.com/vslavik/winsparkle/blob/91bb56ea20eb787c0cbb1f00277f98a4cbef594a/src/settings.h#L181
I'll make a copy of the string to pass to WinSparkle!

Reply:
Fixed in 1532cee!